flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSimpleUDF.java

```diff
@@ -0,0 +1,135 @@
+
+import org.apache.hadoop.hive.ql.exec.FunctionRegistry;
+import org.apache.hadoop.hive.ql.exec.UDF;
+import org.apache.hadoop.hive.ql.exec.UDFArgumentException;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDFUtils;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ *  A ScalarFunction implementation that calls Hive's {@link UDF}.
+ */
+@Internal
+public class HiveSimpleUDF extends HiveScalarFunction<UDF> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(HiveSimpleUDF.class);
+
+	private transient Method method;
+	private transient GenericUDFUtils.ConversionHelper conversionHelper;
+	private transient HiveObjectConversion[] conversions;
+	private transient boolean allIdentityConverter;
+
+	public HiveSimpleUDF(HiveFunctionWrapper<UDF> hiveFunctionWrapper) {
+		super(hiveFunctionWrapper);
+		LOG.info("Creating HiveSimpleUDF from '{}'", this.hiveFunctionWrapper.getClassName());
+	}
+
+	@Override
+	public void openInternal() {
+		LOG.info("Opening HiveSimpleUDF as '{}'", hiveFunctionWrapper.getClassName());
+
+		function = hiveFunctionWrapper.createFunction();
+
+		List<TypeInfo> typeInfos = new ArrayList<>();
+
+		for (DataType arg : argTypes) {
+			typeInfos.add(HiveTypeUtil.toHiveTypeInfo(arg));
+		}
+
+		try {
+			method = function.getResolver().getEvalMethod(typeInfos);
+			returnInspector = ObjectInspectorFactory.getReflectionObjectInspector(method.getGenericReturnType(),
+				ObjectInspectorFactory.ObjectInspectorOptions.JAVA);
+			ObjectInspector[] argInspectors = new ObjectInspector[typeInfos.size()];
+
+			for (int i = 0; i < argTypes.length; i++) {
+				argInspectors[i] = TypeInfoUtils.getStandardJavaObjectInspectorFromTypeInfo(typeInfos.get(i));
+			}
+
+			conversionHelper = new GenericUDFUtils.ConversionHelper(method, argInspectors);
+			conversions = new HiveObjectConversion[argInspectors.length];
+			for (int i = 0; i < argInspectors.length; i++) {
+				conversions[i] = HiveInspectors.getConversion(argInspectors[i]);
+			}
+
+			allIdentityConverter = Arrays.stream(conversions)
+				.allMatch(conv -> conv instanceof IdentityConversion);
+		} catch (Exception e) {
+			throw new FlinkHiveUDFException(
+				String.format("Failed to open HiveSimpleUDF from %s", hiveFunctionWrapper.getClassName()), e);
+		}
+	}
+
+	@Override
+	public Object evalInternal(Object[] args) {
+		checkArgument(args.length == conversions.length);
+
+		if (!allIdentityConverter) {
```

Comment:
How do we gain from this optimization?

Reply:
the flag indicates if all input args can stay the same and don't need to be converted to a Hive object. So if none of them needs to be converted, we can skip the conversion and directly invoke Hive UDF with them.

We can optimize more, if necessary later on, by remembering index of input args that need to be converted and only convert a subset of input args. Currently I'll just make it the same as that in Blink

Reply:
Yeah, but I was wondering how or how much it saves and whether it's worth the effort. It requires all be identityconverter, which is quite limiting in my opinion. Looking at IdentityConverter, the conversion just returns the original input, which is very light.

I'm trying to see if I have missed anything.

Reply:
just by the logic itself, I think it can be faster for complex types and nested types, I don't have benchmarks and users' usage pattern.

I don't think it's critically important, either removing or keeping it. This is more of referencing how we did it in Blink rather than intentionally trying to optimize it hard. If you feel really strong on removing it, we can surely do that

Reply:
It's okay. I was just to see if I missed anything obvious.

Reply:
If there are more than one parameter, this will lead to multiple virtual function calls, and the cost of virtual function calls is not small. This optimization is so simple that I think it's worth doing.